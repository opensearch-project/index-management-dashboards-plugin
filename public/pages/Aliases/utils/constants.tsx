/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { IAlias } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "alias" as keyof IAlias,
  sortDirection: SortDirection.DESC,
  status: "",
  dataSourceId: "",
};
