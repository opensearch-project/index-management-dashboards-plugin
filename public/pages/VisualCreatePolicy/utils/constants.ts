/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AliasAction, AllocationAction } from "../../../../models/interfaces";

export enum ActionType {
  Alias = "alias",
  Allocation = "allocation",
  Close = "close",
  Delete = "delete",
  ForceMerge = "force_merge",
  IndexPriority = "index_priority",
  Notification = "notification",
  Open = "open",
  ReadOnly = "read_only",
  ReadWrite = "read_write",
  ReplicaCount = "replica_count",
  Rollover = "rollover",
  Rollup = "rollup",
  Shrink = "shrink",
  Snapshot = "snapshot",
}

export const DEFAULT_POLICY = {
  description: "A simple default policy that changes the replica count between hot and cold states.",
  default_state: "hot",
  states: [
    {
      name: "hot",
      actions: [{ replica_count: { number_of_replicas: 5 } }],
      transitions: [
        {
          state_name: "cold",
          conditions: { min_index_age: "30d" },
        },
      ],
    },
    {
      name: "cold",
      actions: [{ replica_count: { number_of_replicas: 2 } }],
      transitions: [],
    },
  ],
  ism_template: [],
};

export const EMPTY_DEFAULT_POLICY = {
  description: "A sample description of the policy",
  default_state: "",
  states: [],
  ism_template: [],
};

export const DEFAULT_LEGACY_ERROR_NOTIFICATION = {
  destination: {
    slack: {
      url: "<url>",
    },
  },
  message_template: {
    source: "The index {{ctx.index}} failed during policy execution.",
  },
};

export const DEFAULT_ALIAS: AliasAction = {
  alias: {
    actions: [],
  },
};

export const DEFAULT_ALLOCATION: AllocationAction = {
  allocation: {
    require: {},
    include: {},
    exclude: {},
    wait_for: false,
  },
};

export const DEFAULT_CLOSE = {
  close: {},
};
export const DEFAULT_DELETE = {
  delete: {},
};
export const DEFAULT_FORCE_MERGE = {
  force_merge: {
    max_num_segments: 1,
  },
};
export const DEFAULT_INDEX_PRIORITY = {
  index_priority: {
    priority: 1,
  },
};
export const DEFAULT_NOTIFICATION = {
  notification: {
    channel: {
      id: "",
    },
    message_template: {
      source: "",
    },
  },
};
export const DEFAULT_OPEN = {
  open: {},
};
export const DEFAULT_READ_ONLY = {
  read_only: {},
};
export const DEFAULT_READ_WRITE = {
  read_write: {},
};
export const DEFAULT_REPLICA_COUNT = {
  replica_count: {
    number_of_replicas: 1,
  },
};
export const DEFAULT_ROLLOVER = {
  rollover: {
    min_doc_count: 5,
  },
};
export const DEFAULT_ROLLUP = {
  // TODO: In future make actual UI
  rollup: {
    jsonString: JSON.stringify(
      {
        ism_rollup: {
          target_index: "rollup-nyc-taxi-data",
          description: "Example rollup job",
          page_size: 200,
          dimensions: [
            {
              date_histogram: {
                source_field: "tpep_pickup_datetime",
                fixed_interval: "1h",
                timezone: "America/Los_Angeles",
              },
            },
            {
              terms: {
                source_field: "PULocationID",
              },
            },
          ],
          metrics: [
            {
              source_field: "passenger_count",
              metrics: [
                {
                  avg: {},
                },
                {
                  sum: {},
                },
                {
                  max: {},
                },
                {
                  min: {},
                },
                {
                  value_count: {},
                },
              ],
            },
          ],
        },
      },
      null,
      4
    ),
  },
};

export const DEFAULT_SHRINK = {
  shrink: {
    percentage_of_source_shards: 0.5,
    force_unsafe: false,
    target_index_name_template: {
      source: "{{ctx.index}}_shrunken",
    },
  },
  force_unsafe_input: "no",
};

export const DEFAULT_SNAPSHOT = {
  snapshot: {
    repository: "example-repository",
    snapshot: "example-snapshot",
  },
};

export const actions = [
  DEFAULT_ALIAS,
  DEFAULT_ALLOCATION,
  DEFAULT_CLOSE,
  DEFAULT_DELETE,
  DEFAULT_FORCE_MERGE,
  DEFAULT_INDEX_PRIORITY,
  DEFAULT_NOTIFICATION,
  DEFAULT_OPEN,
  DEFAULT_READ_ONLY,
  DEFAULT_READ_WRITE,
  DEFAULT_REPLICA_COUNT,
  DEFAULT_ROLLOVER,
  DEFAULT_ROLLUP,
  DEFAULT_SHRINK,
  DEFAULT_SNAPSHOT,
];

export const ISM_TEMPLATE_INPUT_MAX_WIDTH = "400px";
