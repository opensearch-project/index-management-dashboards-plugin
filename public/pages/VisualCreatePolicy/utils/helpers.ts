/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Action, ISMTemplate, Policy, State, Transition, UIAction } from "../../../../models/interfaces";
import {
  ActionType,
  DEFAULT_ALIAS,
  DEFAULT_ALLOCATION,
  DEFAULT_CLOSE,
  DEFAULT_DELETE,
  DEFAULT_FORCE_MERGE,
  DEFAULT_INDEX_PRIORITY,
  DEFAULT_NOTIFICATION,
  DEFAULT_OPEN,
  DEFAULT_READ_ONLY,
  DEFAULT_READ_WRITE,
  DEFAULT_REPLICA_COUNT,
  DEFAULT_ROLLOVER,
  DEFAULT_ROLLUP,
  DEFAULT_SHRINK,
  DEFAULT_SNAPSHOT,
} from "./constants";
import {
  AliasUIAction,
  AllocationUIAction,
  CloseUIAction,
  DeleteUIAction,
  ForceMergeUIAction,
  IndexPriorityUIAction,
  NotificationUIAction,
  OpenUIAction,
  ReadOnlyUIAction,
  ReadWriteUIAction,
  ReplicaCountUIAction,
  RolloverUIAction,
  RollupUIAction,
  ShrinkUIAction,
  SnapshotUIAction,
} from "../components/UIActions";

export const getConditionContent = (transition: Transition): string => {
  const {
    conditions: {
      min_doc_count: minDocCount = undefined,
      min_index_age: minIndexAge = undefined,
      min_size: minSize = undefined,
      min_rollover_age: minRolloverAge = undefined,
      cron = undefined,
    } = {},
  } = transition;
  if (minSize != undefined) return `Minimum index size is ${minSize}`;
  if (minDocCount != undefined) return `Minimum index doc count is ${minDocCount}`;
  if (minIndexAge != undefined) return `Minimum index age is ${minIndexAge}`;
  if (minRolloverAge != undefined) return `Minimum rollover age is ${minRolloverAge}`;
  if (cron != undefined) return `After cron expression "${cron.cron.expression}" in ${cron.cron.timezone}`;
  return "No condition";
};

export const getUIActionFromData = (action: Action): UIAction<any> => {
  const actionType = Object.keys(action)
    .filter((key) => key !== "timeout" && key !== "retry")
    .pop();
  if (!actionType) throw new Error(`Failed to get action using type [${actionType}]`);
  const uiAction = getUIAction(actionType);
  return uiAction.clone(action);
};

export const getUIAction = (actionType: string): UIAction<any> => {
  switch (actionType) {
    case ActionType.Alias:
      return new AliasUIAction(DEFAULT_ALIAS);
    case ActionType.Allocation:
      return new AllocationUIAction(DEFAULT_ALLOCATION);
    case ActionType.Close:
      return new CloseUIAction(DEFAULT_CLOSE);
    case ActionType.Delete:
      return new DeleteUIAction(DEFAULT_DELETE);
    case ActionType.ForceMerge:
      return new ForceMergeUIAction(DEFAULT_FORCE_MERGE);
    case ActionType.IndexPriority:
      return new IndexPriorityUIAction(DEFAULT_INDEX_PRIORITY);
    case ActionType.Notification:
      return new NotificationUIAction(DEFAULT_NOTIFICATION);
    case ActionType.Open:
      return new OpenUIAction(DEFAULT_OPEN);
    case ActionType.ReadOnly:
      return new ReadOnlyUIAction(DEFAULT_READ_ONLY);
    case ActionType.ReadWrite:
      return new ReadWriteUIAction(DEFAULT_READ_WRITE);
    case ActionType.ReplicaCount:
      return new ReplicaCountUIAction(DEFAULT_REPLICA_COUNT);
    case ActionType.Rollover:
      return new RolloverUIAction(DEFAULT_ROLLOVER);
    case ActionType.Rollup:
      return new RollupUIAction(DEFAULT_ROLLUP);
    case ActionType.Shrink:
      return new ShrinkUIAction(DEFAULT_SHRINK);
    case ActionType.Snapshot:
      return new SnapshotUIAction(DEFAULT_SNAPSHOT);
    default:
      throw new Error(`Action type [${actionType}] not supported`);
  }
};

export const getActionOptions = (actionRepoSingleton: ActionRepository) => {
  return actionRepoSingleton.getAllActionTypes().map((key) => {
    let uiAction;
    let text = key
      .split("_")
      .map((str) => capitalizeFirstLetter(str))
      .join(" ");
    try {
      uiAction = getUIAction(key)?.customDisplayText;
      if (uiAction !== undefined) text = uiAction;
    } catch (e) {}
    return {
      value: key,
      text: text,
    };
  });
};

class ActionRepository {
  repository: { [actionType: string]: [new (action: Action) => UIAction<any>, Action] } = {
    alias: [AliasUIAction, DEFAULT_ALIAS],
    allocation: [AllocationUIAction, DEFAULT_ALLOCATION],
    close: [CloseUIAction, DEFAULT_CLOSE],
    delete: [DeleteUIAction, DEFAULT_DELETE],
    force_merge: [ForceMergeUIAction, DEFAULT_FORCE_MERGE],
    index_priority: [IndexPriorityUIAction, DEFAULT_INDEX_PRIORITY],
    notification: [NotificationUIAction, DEFAULT_NOTIFICATION],
    open: [OpenUIAction, DEFAULT_OPEN],
    read_only: [ReadOnlyUIAction, DEFAULT_READ_ONLY],
    read_write: [ReadWriteUIAction, DEFAULT_READ_WRITE],
    replica_count: [ReplicaCountUIAction, DEFAULT_REPLICA_COUNT],
    rollover: [RolloverUIAction, DEFAULT_ROLLOVER],
    rollup: [RollupUIAction, DEFAULT_ROLLUP],
    shrink: [ShrinkUIAction, DEFAULT_SHRINK],
    snapshot: [SnapshotUIAction, DEFAULT_SNAPSHOT],
  };

  getAllActionTypes = () => {
    return Object.keys(this.repository);
  };

  registerAction = (actionType: string, uiActionCtor: new (action: Action) => UIAction<any>, defaultAction: Action) => {
    if (this.repository.hasOwnProperty(actionType)) {
      throw new Error(`Cannot register an action twice in the repository [type=${actionType}]`);
    }

    this.repository[actionType] = [uiActionCtor, defaultAction];
  };

  getUIActionFromData = (action: Action) => {
    const actionType = Object.keys(action)
      .filter((key) => key !== "timeout" && key !== "retry")
      .pop();
    if (!actionType) throw new Error(`Failed to get action using type [${actionType}]`);
    const uiAction = this.getUIAction(actionType);
    return uiAction.clone(action);
  };

  getUIAction = (actionType: string): UIAction<any> => {
    const uiAction = this.repository[actionType];
    if (!uiAction) throw new Error(`Action type [${actionType}] not supported`);
    return new uiAction[0](uiAction[1]);
  };
}

export const actionRepoSingleton = new ActionRepository();

// Takes in the ismTemplates which could be a single object, array, or not defined and returns them reformatted as a list
export const convertTemplatesToArray = (ismTemplates: ISMTemplate[] | ISMTemplate | null | undefined): ISMTemplate[] => {
  const templates = [];
  // policy.ism_template can be an array of templates or a single template as an object or null
  if (Array.isArray(ismTemplates)) {
    templates.push(...ismTemplates);
  } else if (ismTemplates) {
    templates.push(ismTemplates);
  }
  return templates;
};

export const capitalizeFirstLetter = ([first, ...rest]: string, locale = navigator.language) =>
  first.toLocaleUpperCase(locale) + rest.join("");

export const getOrderInfo = (
  state: State | null,
  states: State[]
): { order: string; afterBeforeState: string; disableOrderSelections: boolean } => {
  const isEditing = !!state;
  const editingStateIdx = states.findIndex((s) => s === state);
  let afterBeforeState = "";
  let order = "after";
  let disableOrderSelections = false;
  if (isEditing) {
    // If there is only one existing state then we are editing the only state and there is nothing to order after/before
    if (states.length === 1) {
      disableOrderSelections = true;
    } else {
      // If the editing state is the first state then we are "Add before $secondState"
      if (editingStateIdx === 0) {
        afterBeforeState = states[editingStateIdx + 1]?.name;
        order = "before";
      } else {
        // otherwise we are "Add after $previousState"
        afterBeforeState = states[editingStateIdx - 1]?.name;
      }
    }
  } else {
    // when creating a new state
    // If there are no existing states then there is nothing to order after/before
    if (!states.length) {
      disableOrderSelections = true;
    } else {
      // Else always order it after the last state
      afterBeforeState = states[states.length - 1]?.name;
    }
  }

  return { order, afterBeforeState, disableOrderSelections };
};

export const getUpdatedStates = (
  state: State,
  editingState: State | null,
  states: State[],
  order: string,
  afterBeforeState: string
): State[] => {
  const someStates: State[] = [];
  const isEditing = !!editingState;
  if ((isEditing && states.length === 1) || (!isEditing && !states.length)) return [state];

  states.forEach((s, idx) => {
    if (s.name === afterBeforeState && order === "before") someStates.push(state);
    if (s.name !== editingState?.name) someStates.push(s);
    if (s.name === afterBeforeState && order === "after") someStates.push(state);
  });

  return someStates;
};

export const getUpdatedPolicy = (
  currentPolicy: Policy,
  updatedState: State,
  editingState: State | null,
  currentStates: State[],
  order: string,
  afterBeforeState: string
): Policy => {
  const updatedStates = getUpdatedStates(updatedState, editingState, currentStates, order, afterBeforeState);
  let defaultState = currentPolicy.default_state;
  // If there is 1 state, change it to this state
  if (updatedStates.length === 1) {
    defaultState = updatedStates[0].name;
  }
  // Change the default state if the state being edited was the default state
  if (editingState && editingState.name === defaultState) {
    // don't bother checking if the name itself changed, just set it regardless to the value from the new state
    defaultState = updatedState.name;
  }

  return {
    ...currentPolicy,
    states: updatedStates,
    default_state: defaultState,
  };
};
