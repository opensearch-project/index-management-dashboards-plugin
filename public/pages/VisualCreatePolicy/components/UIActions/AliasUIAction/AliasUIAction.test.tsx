/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DEFAULT_ALIAS } from "../../../utils/constants";
import { AliasAction, AliasActions, UIAction } from "../../../../../../models/interfaces";
import { actionRepoSingleton } from "../../../utils/helpers";

const TEST_PROPS: UIAction<AliasAction> = { action: DEFAULT_ALIAS } as UIAction<AliasAction>;

const renderComponent = (uiAction: UIAction<AliasAction> = TEST_PROPS) => {
  render(actionRepoSingleton.getUIAction("alias").render(uiAction, mockOnChangeAction));
};
const mockOnChangeAction = (uiAction: UIAction<AliasAction> = TEST_PROPS) => {
  cleanup();
  renderComponent(uiAction);
};

afterEach(() => cleanup());

describe("AliasUIAction component", () => {
  it("renders with blank action", () => {
    const { container } = render(actionRepoSingleton.getUIAction("alias").render(TEST_PROPS, mockOnChangeAction));

    userEvent.click(screen.getByTestId("add-alias-toggle"));
    userEvent.click(screen.getByTestId("remove-alias-toggle"));

    const addAliasRow = screen.getByTestId("add-alias-row");
    const removeAliasRow = screen.getByTestId("remove-alias-row");

    expect(addAliasRow).toHaveTextContent("Select aliases to add to indexes");
    expect(addAliasRow).toHaveTextContent("You can add up to 10 more aliases.");

    expect(removeAliasRow).toHaveTextContent("Select aliases to remove from indexes");
    expect(removeAliasRow).toHaveTextContent("You can add up to 10 more aliases.");

    expect(container).toMatchSnapshot();
  });

  it("renders with pre-defined actions", () => {
    const actions = [
      { [AliasActions.ADD]: { alias: "alias1" } },
      { [AliasActions.ADD]: { alias: "alias3" } },
      { [AliasActions.REMOVE]: { alias: "alias2" } },
      { [AliasActions.ADD]: { aliases: ["alias5", "alias7", "alias9"] } },
      { [AliasActions.REMOVE]: { aliases: ["alias4", "alias6", "alias8"] } },
    ];
    const testProps = { ...TEST_PROPS };
    testProps.action.alias.actions = actions;

    const { container } = render(actionRepoSingleton.getUIAction("alias").render(testProps, mockOnChangeAction));
    expect(screen.getByTestId("add-alias-row")).toHaveTextContent("You can add up to 5 more aliases.");
    expect(screen.getByTestId("remove-alias-row")).toHaveTextContent("You can add up to 6 more aliases.");

    const addAliasComboBox = screen.getByTestId("add-alias-combo-box");
    const removeAliasComboBox = screen.getByTestId("remove-alias-combo-box");
    actions.forEach((action) => {
      switch (Object.keys(action)[0]) {
        case AliasActions.ADD:
          if (action[AliasActions.ADD].alias) expect(addAliasComboBox).toHaveTextContent(action[AliasActions.ADD].alias);
          if (action[AliasActions.ADD].aliases)
            action[AliasActions.ADD].aliases.forEach((alias) => {
              expect(addAliasComboBox).toHaveTextContent(alias);
            });
          break;
        case AliasActions.REMOVE:
          if (action[AliasActions.REMOVE].alias) expect(removeAliasComboBox).toHaveTextContent(action[AliasActions.REMOVE].alias);
          if (action[AliasActions.REMOVE].aliases)
            action[AliasActions.REMOVE].aliases.forEach((alias) => {
              expect(removeAliasComboBox).toHaveTextContent(alias);
            });
          break;
      }
    });

    expect(container).toMatchSnapshot();
  });
});
