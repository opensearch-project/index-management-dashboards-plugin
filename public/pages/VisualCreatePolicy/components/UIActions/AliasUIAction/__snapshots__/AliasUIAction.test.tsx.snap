// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AliasUIAction component renders with blank action 1`] = `
<div>
  <div
    class="euiSwitch euiSwitch--primary"
  >
    <button
      aria-checked="true"
      aria-labelledby="some_html_id"
      class="euiSwitch__button"
      data-test-subj="add-alias-toggle"
      id="some_html_id"
      role="switch"
      type="button"
    >
      <span
        class="euiSwitch__body"
      >
        <span
          class="euiSwitch__thumb"
        />
        <span
          class="euiSwitch__track"
        >
          EuiIconMock
          EuiIconMock
        </span>
      </span>
    </button>
    <span
      class="euiSwitch__label"
      id="some_html_id"
    >
      Add aliases
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--m"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth"
    data-test-subj="add-alias-row"
    id="some_html_id-row"
    style="max-width: 100%;"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="some_html_id"
      >
        Select aliases to add to indexes
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        data-test-subj="add-alias-combo-box"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <p
                class="euiComboBoxPlaceholder"
              >
                Enter aliases to add
              </p>
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiText euiText--extraSmall"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          You can add up to 10 more aliases.
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiSwitch euiSwitch--primary"
  >
    <button
      aria-checked="true"
      aria-labelledby="some_html_id"
      class="euiSwitch__button"
      data-test-subj="remove-alias-toggle"
      id="some_html_id"
      role="switch"
      type="button"
    >
      <span
        class="euiSwitch__body"
      >
        <span
          class="euiSwitch__thumb"
        />
        <span
          class="euiSwitch__track"
        >
          EuiIconMock
          EuiIconMock
        </span>
      </span>
    </button>
    <span
      class="euiSwitch__label"
      id="some_html_id"
    >
      Remove aliases
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--m"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth"
    data-test-subj="remove-alias-row"
    id="some_html_id-row"
    style="max-width: 100%;"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="some_html_id"
      >
        Select aliases to remove from indexes
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        data-test-subj="remove-alias-combo-box"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <p
                class="euiComboBoxPlaceholder"
              >
                Enter aliases to remove
              </p>
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiText euiText--extraSmall"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          You can add up to 10 more aliases.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AliasUIAction component renders with pre-defined actions 1`] = `
<div>
  <div
    class="euiSwitch euiSwitch--primary"
  >
    <button
      aria-checked="true"
      aria-labelledby="some_html_id"
      class="euiSwitch__button"
      data-test-subj="add-alias-toggle"
      id="some_html_id"
      role="switch"
      type="button"
    >
      <span
        class="euiSwitch__body"
      >
        <span
          class="euiSwitch__thumb"
        />
        <span
          class="euiSwitch__track"
        >
          EuiIconMock
          EuiIconMock
        </span>
      </span>
    </button>
    <span
      class="euiSwitch__label"
      id="some_html_id"
    >
      Add aliases
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--m"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth"
    data-test-subj="add-alias-row"
    id="some_html_id-row"
    style="max-width: 100%;"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="some_html_id"
      >
        Select aliases to add to indexes
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        data-test-subj="add-alias-combo-box"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias1"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias1
                  </span>
                  <button
                    aria-label="Remove alias1 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias1 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias3"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias3
                  </span>
                  <button
                    aria-label="Remove alias3 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias3 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias5"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias5
                  </span>
                  <button
                    aria-label="Remove alias5 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias5 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias7"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias7
                  </span>
                  <button
                    aria-label="Remove alias7 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias7 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias9"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias9
                  </span>
                  <button
                    aria-label="Remove alias9 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias9 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
            <div
              class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
            >
              <button
                aria-label="Clear input"
                class="euiFormControlLayoutClearButton"
                data-test-subj="comboBoxClearButton"
                type="button"
              >
                EuiIconMock
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiText euiText--extraSmall"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          You can add up to 5 more aliases.
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiSwitch euiSwitch--primary"
  >
    <button
      aria-checked="true"
      aria-labelledby="some_html_id"
      class="euiSwitch__button"
      data-test-subj="remove-alias-toggle"
      id="some_html_id"
      role="switch"
      type="button"
    >
      <span
        class="euiSwitch__body"
      >
        <span
          class="euiSwitch__thumb"
        />
        <span
          class="euiSwitch__track"
        >
          EuiIconMock
          EuiIconMock
        </span>
      </span>
    </button>
    <span
      class="euiSwitch__label"
      id="some_html_id"
    >
      Remove aliases
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--m"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth"
    data-test-subj="remove-alias-row"
    id="some_html_id-row"
    style="max-width: 100%;"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="some_html_id"
      >
        Select aliases to remove from indexes
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        data-test-subj="remove-alias-combo-box"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias2"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias2
                  </span>
                  <button
                    aria-label="Remove alias2 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias2 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias4"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias4
                  </span>
                  <button
                    aria-label="Remove alias4 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias4 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias6"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias6
                  </span>
                  <button
                    aria-label="Remove alias6 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias6 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <span
                class="euiBadge euiBadge--hollow euiBadge--iconRight euiComboBoxPill"
                title="alias8"
              >
                <span
                  class="euiBadge__content"
                >
                  <span
                    class="euiBadge__text"
                  >
                    alias8
                  </span>
                  <button
                    aria-label="Remove alias8 from selection in this group"
                    class="euiBadge__iconButton"
                    title="Remove alias8 from selection in this group"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </span>
              </span>
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
            <div
              class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
            >
              <button
                aria-label="Clear input"
                class="euiFormControlLayoutClearButton"
                data-test-subj="comboBoxClearButton"
                type="button"
              >
                EuiIconMock
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiText euiText--extraSmall"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          You can add up to 6 more aliases.
        </div>
      </div>
    </div>
  </div>
</div>
`;
